sql/core/src/main/scala/org/apache/spark/sql/execution/Generate.scala

```diff
@@ -74,10 +84,15 @@ case class Generate(
           } else {
             outputRows.map(or => joinProjection(joinedRow(row, or)))
           }
+        } ++ LazyIterator(() => boundGenerator.terminate()).map { row =>
+          joinedRow(nullOriginalRow, row)
         }
       }
     } else {
-      child.execute().mapPartitions(iter => iter.flatMap(row => boundGenerator.eval(row)))
+      child.execute().mapPartitions(iter =>
+        iter.flatMap(row => boundGenerator.eval(row)) ++
+          LazyIterator(() => boundGenerator.terminate())
```

Comment:
Yes, I that's OK by called in each partition.

See 
https://github.com/apache/hive/blob/trunk/ql/src/java/org/apache/hadoop/hive/ql/exec/mr/ExecReducer.java#L278
https://github.com/apache/hive/blob/trunk/ql/src/java/org/apache/hadoop/hive/ql/exec/mr/ExecMapper.java#L192
The `Operator.close()` is called in `MapReduceBase.close`, which mean they are supposed to run once within each Mapper/Reducer.

And 
https://github.com/apache/hive/blob/trunk/ql/src/java/org/apache/hadoop/hive/ql/exec/UDTFOperator.java#L144
https://github.com/apache/hive/blob/trunk/ql/src/java/org/apache/hadoop/hive/ql/exec/Operator.java#L616
shows the `genericUDTF.close()` is called within the `Operator.close`.

Sorry, please correct me if my understanding is wrong.